common/src/main/java/com/microsoft/identity/common/internal/controllers/BaseController.java

```diff
@@ -375,6 +377,135 @@ protected void renewAccessToken(@NonNull final SilentTokenCommandParameters para
                                     + isRemoved
                     );
                 }
+
+                /*
+                    Intended to cover the AAD outage scenario for the refresh_in logic.
+                    Should return existing AT without refreshing it.
+                    This way caller will know whether to refresh based on this exception.
+                 */
+                if(SERVICE_NOT_AVAILABLE.equals(errorCode)){
+                    throw new ServiceException(SERVICE_NOT_AVAILABLE, "AAD is not available.", tokenResult.getErrorResponse().getStatusCode(), null);
+                }
+
+            } else {
+                Logger.warn(TAG, "Invalid state, No token success or error response on the token result");
+            }
+        }
+    }
+
+    /*
+        Light version of:
+        protected renewAccessToken(@NonNull final SilentTokenCommandParameters parameters,
+                                   @NonNull final AcquireTokenResult acquireTokenSilentResult,
+                                   @SuppressWarnings(WarningType.rawtype_warning) @NonNull final OAuth2TokenCache tokenCache,
+                                   @SuppressWarnings(WarningType.rawtype_warning) @NonNull final OAuth2Strategy strategy,
+                                   @NonNull final ICacheRecord cacheRecord)
+
+        Diffs
+        1) acquireTokenSilentResult updates omitted.
+        2) All arguments derived from SilentTokenCommandParameters
+        3) New logic replacing old Access Token
+     */
+    public void renewAccessToken(@NonNull final SilentTokenCommandParameters parameters)
```

Comment:
This should be an `abstract` method in `BaseController` and then we can overwrite it in specific controllers..because otherwise I don't know how we would do this for Joined Accounts...my understanding is that the `JoinedAccountController` will Override this method to do the refresh using the PRT

Reply:
Also let's make the input specific to the refresh on logic...you don't need to pass around the entire SilentTokenCommand  object...you can just grab the necessary information out of it and supply it to a custom parameter object created for your command and that way you don't have all this similar looking code in a lot of places.

Reply:
There is a tradeoff right? We can make it abstract, but we should make the reusable operations available from the BaseController. Maybe we can rename this method to mean for non-joined accounts, but the renewal operation is the same for at least 2 controllers in this case.

Reply:
Yes, I agree. Let's keep as is then.